tests/utils.py

```diff
@@ -74,3 +77,118 @@ def wait_for_server(host, port, timeout=10):
+
+      popen_cwd:
+        Current working directory used when instancing a
+        subprocess.Popen object.
+        Default is "."
+
+      extra_cmd_args:
+        List of additional arguments for the command
+        which will be ran from the subprocess.
+        More precisely "python -u <path_to_server> <port> <extra_cmd_args>".
+        Default is empty list.
+  """
+
+  # Starts subprocess which runs the server,
+  # sets up the TempFile, waits for server response.
+  def __init__(self, log, server='simple_server.py',
+      port=None, timeout=10, popen_cwd=".",
+      extra_cmd_args=[]):
+    # Create temporary log file used for logging stdout and stderr
+    # of the subprocess. In the mode "r+"" stands for reading and writing
+    # and "t" stands for text mode.
+    self.__temp_log_file = tempfile.TemporaryFile(mode='r+t')
+
+    self.__SERVER_PORT = port or random.randint(30000, 45000)
+    self.__logger = log
+
+    # The "-u" option forces stdin, stdout and stderr to be unbuffered.
+    command = ['python', '-u', server, str(self.__SERVER_PORT)] \
+        + extra_cmd_args
+
+    # We are reusing one server subprocess in multiple unit tests, but we are
+    # collecting the logs per test.
+    self.__server_process = subprocess.Popen(command,
+        stdout=self.__temp_log_file, stderr=subprocess.STDOUT, cwd=popen_cwd)
+
+    self.__logger.info('Server process started.')
+    self.__logger.info('Server process id: ' + str(self.__server_process.pid))
+    self.__logger.info('Serving on port: ' + str(self.__SERVER_PORT))
+
+    if timeout > 0:
+      try:
+        wait_for_server('localhost', self.__SERVER_PORT, timeout)
+      except Exception as e:
+        self.__logger.error(
+            "Error while waiting for the server to start: " + repr(e))
+        # Make sure that errors from the server side will be logged.
+        self.flush_log()
+        raise e
+
+
+  @property
+  def port(self):
+    return self.__SERVER_PORT
+
+
+  # Logs contents from TempFile, truncates buffer
+  def flush_log(self, test_name=""):
+
+    # Seek is needed to move the pointer to the beginning of the file, because
+    # the subprocess could have read and/or write and thus moved the pointer.
+    self.__temp_log_file.seek(0)
+    log_message = self.__temp_log_file.read()
+
+    if len(log_message) > 0:
+      self.__logger.info("*******************")
+      # If you run only one test file with "python <file_name>"
+      # you will receive unnecessary "(null)" and "__main__."
+      test_name = test_name.replace("__main__.", "")
+      log_message = log_message.replace("(null): ", "")
+
+      self.__logger.info("Logs from the server subprocess!")
+      if test_name:
+        self.__logger.info("Logs for test " + test_name)
+
+      self.__logger.info("INFO: " + log_message)
+      self.__logger.info("*******************")
+      # Make sure the file is empty before the next test logs new information.
+      self.__temp_log_file.truncate(0)
+
```

Comment:
> We don't print test names in any other logging either: if it's not clear which test is running, why not improve that in general instead of just for server output

I think the `test_name` argument can be removed because as we discussed on a test failure the logs are showed once before all tests and a second time just after the test name:
![image](https://user-images.githubusercontent.com/16246778/92588350-86983580-f2a1-11ea-9537-76de15931b4f.png)

I don't think we can add the test names to all logs because many of the logs are coming from public API functions and there is no need to add `test_name` argument just for that.

> if the log message from the server does not look good, I think we should fix the test server output instead of replacing things in the output here

Well, yes, maybe you are right. I will remove that part.

> the output missing the server name however: e.g. "simple_server.py": this would be important sometimes

Will add that.